process/block/shardblock.go

```diff
@@ -1349,7 +1339,6 @@ func (sp *shardProcessor) receivedMetaBlock(headerHandler data.HeaderHandler, me
 		return
 	}
 
-	sp.epochStartTrigger.ReceivedHeader(metaBlock)
 	if sp.epochStartTrigger.IsEpochStart() {
```

Comment:
Here we can have a deadlock, as this channel "chRcvEpochStart" is not buffered and it is consumed only in the method requestEpochStartInfo called from ProcessBlock. This means that if node will not reach to that point all the metablocks received meantime will be stuck in this point, as now you will not call the sp.epochStartTrigger.ReceivedHeader(metaBlock) method synchronous, from here, anymore.
I will suggest to remove this channel at all and also these 3 lines from here, and to change the method requestEpochStartInfo like this:

func (sp *shardProcessor) requestEpochStartInfo(header *block.Header, haveTime func() time.Duration) error {
	haveMissingMetaHeaders := header.IsStartOfEpochBlock() && !sp.epochStartTrigger.IsEpochStart()
	if !haveMissingMetaHeaders {
		return nil
	}

	for {
		time.Sleep(time.Millisecond)
		
		if haveTime() < 0 {
			break
		}
		
		if sp.epochStartTrigger.IsEpochStart() {
			return nil
		}
	}

	return process.ErrTimeIsOut
}

Reply:
As an alternative, as to not have a blocking write, @sasurobert could have written something like:
```GO
select{
    case sp.chRcvEpochStart <- true:
    default:
}
```